src/pyroboplan/trajectory/trajectory_optimization.py

```diff
@@ -0,0 +1,423 @@
+            k : int
+                The trajectory segment index.
+            n : int
+                The degree of freedom index.
+            step : float
+                The normalized distance, between 0 and 1, along that segment.
+
+        Return
+        ------
+            pydrake.autodiffutils.AutodiffXd
+                The jerk along the specified segment evaluated at the specified step.
+        """
+        return 8.0 * (x_d[k, n] - 2.0 * xc_d[k, n] + x_d[k + 1, n]) / h[k] ** 2
+
+    def plan(self, q_path):
+        """
+        Plans a trajectory from a start to a goal configuration, or along an entire trajectory.
+
+        If the input list has 2 elements, then this is assumed to be the start and goal configurations.
+        The intermediate waypoints will be determined automatically.
+
+        If the input list has more than 2 elements, then these are the actual waypoints that must be achieved.
+
+        Parameters
+        ----------
+            q_path : list[array-like]
+                A list of joint configurations describing the desired motion.
+
+        Return
+        ------
+            Optional[pyroboplan.trajectory.polynomial.CubicPolynomialTrajectory]
+                The resulting trajectory, or None if optimization failed
+        """
+        if len(q_path) == 0:
+            warnings.warn("Cannot optimize over an empty path.")
+            return None
+        num_waypoints = self.options.num_waypoints
+        num_dofs = len(q_path[0])
+
+        if len(q_path) == num_waypoints:
+            fully_specified_path = True
+        elif len(q_path) == 2:
+            fully_specified_path = False
+        else:
+            raise ValueError("Path must either be length 2 or equal to num_waypoints.")
+
+        # Preprocess the kinematic limits
+        min_vel = self._process_limits(self.options.min_vel, num_dofs, "min_vel")
+        max_vel = self._process_limits(self.options.max_vel, num_dofs, "max_vel")
+        min_accel = self._process_limits(self.options.min_accel, num_dofs, "min_accel")
+        max_accel = self._process_limits(self.options.max_accel, num_dofs, "max_accel")
+        min_jerk = self._process_limits(self.options.min_jerk, num_dofs, "min_jerk")
+        max_jerk = self._process_limits(self.options.max_jerk, num_dofs, "max_jerk")
+
+        # Initialize the basic program and its variables
+        prog = MathematicalProgram()
+        x = prog.NewContinuousVariables(num_waypoints, num_dofs)
+        x_d = prog.NewContinuousVariables(num_waypoints, num_dofs)
+        xc = prog.NewContinuousVariables(num_waypoints - 1, num_dofs)
+        xc_d = prog.NewContinuousVariables(num_waypoints - 1, num_dofs)
+        h = prog.NewContinuousVariables(num_waypoints - 1)
+
+        # Initial and final conditions
+        if fully_specified_path:
+            for idx in range(num_waypoints):
+                prog.AddBoundingBoxConstraint(q_path[idx], q_path[idx], x[idx, :])
```

Comment:
technically this is not an initial or final condition, so might be good to revise the comment above